lib/ABI/FunctionType.cpp

```diff
@@ -613,12 +650,98 @@ convertToCABIFunctionType(const model::RawFunctionType &Function,
     return ConvertionHelper<A>::toRaw(Function, TheBinary);
   });
 }
 
-} // namespace model
+Layout layout(const model::CABIFunctionType &Function) {
+  auto R = skippingEnumSwitch<1>(Function.ABI, [&]<model::ABI::Values A>() {
+    return ConvertionHelper<A>::getLayout(Function);
+  });
+  return Layout(std::move(R.Arguments),
+                std::move(R.ReturnValue),
+                std::move(R.CalleeSavedRegisters),
+                std::move(R.FinalStackOffset));
+}
+
+Layout layout(const model::RawFunctionType &Function) {
+  // Lay register arguments out.
+  std::decay<decltype(Layout::Arguments)>::type ResultArguments;
+  for (const model::NamedTypedRegister &Register : Function.Arguments)
+    ResultArguments.emplace_back().Registers = { Register.Location };
+
+  // Lay stack arguments out.
+  const model::Type *OriginalStackType = Function.StackArgumentsType.get();
+  auto *StackStruct = llvm::dyn_cast<model::StructType>(OriginalStackType);
+  revng_assert(StackStruct,
+               "`RawFunctionType::StackArgumentsType` must be a struct.");
+  typename Layout::Argument::StackSpan StackSpan{ 0, StackStruct->Size };
+  ResultArguments.emplace_back().Stack = std::move(StackSpan);
+
+  // Lay the return value out.
+  std::decay<decltype(Layout::ReturnValue)>::type ResultReturnValue;
+  for (const model::TypedRegister &Register : Function.ReturnValues)
+    ResultReturnValue.Registers.emplace_back(Register.Location);
+
+  // Fill callee saved registers.
+  std::decay<decltype(Layout::CalleeSavedRegisters)>::type ResultCSRegisters;
+  ResultCSRegisters.resize(Function.PreservedRegisters.size());
+  llvm::copy(Function.PreservedRegisters, ResultCSRegisters.begin());
+
+  return Layout(std::move(ResultArguments),
+                std::move(ResultReturnValue),
+                std::move(ResultCSRegisters),
+                Function.FinalStackOffset);
+}
+
+bool Layout::verify() const {
+  model::Architecture::Values ExpectedArch = model::Architecture::Invalid;
+  std::unordered_set<model::Register::Values> LookupHelper;
+  auto VerificationHelper = [&](model::Register::Values Register) -> bool {
+    if (!LookupHelper.emplace(Register).second)
+      return false;
+    if (ExpectedArch == model::Architecture::Invalid)
+      ExpectedArch = model::Register::getArchitecture(Register);
+    else if (ExpectedArch != model::Register::getArchitecture(Register))
+      return false;
+
+    return true;
+  };
+
+  // Verify arguments
+  LookupHelper.clear();
+  for (const Layout::Argument &Argument : Arguments)
+    for (model::Register::Values Register : Argument.Registers)
+      if (!VerificationHelper(Register))
+        return false;
+
+  // Verify return values
+  LookupHelper.clear();
+  for (model::Register::Values Register : ReturnValue.Registers)
+    if (!VerificationHelper(Register))
+      return false;
+
+  if (CalleeSavedRegisters.empty())
+    return false;
```

Comment:
Why?

Reply:
I can't imagine an ABI without at least one of those. Might be a good indication that the layout is not doing right???

No big deal thought, I'm fine with getting rid of it.

Reply:
Mmh, I'm thinking the user might to something stupid with `RawFunctionType` (no specific ABI involved).
Up to you.

Reply:
Gone: https://github.com/revng/revng/pull/131/commits/0bc0bc7396467e8916fd0f908d4c77ddbdf198fc.